rclcpp_action/src/server.cpp

```diff
@@ -12,30 +12,474 @@
+    auto iter = pimpl_->result_requests_.find(uuid);
+    if (iter == pimpl_->result_requests_.end()) {
+      pimpl_->result_requests_[uuid] = std::vector<rmw_request_id_t>{request_header};
+    } else {
+      iter->second.push_back(request_header);
+    }
+  }
+}
+
+void
+ServerBase::execute_check_expired_goals()
+{
+  // Allocate expecting only one goal to expire at a time
+  rcl_action_goal_info_t expired_goals[1];
+  size_t num_expired = 1;
+
+  // Loop in case more than 1 goal expired
+  while (num_expired > 0u) {
+    rcl_ret_t ret = rcl_action_expire_goals(
+      pimpl_->action_server_.get(), expired_goals, 1, &num_expired);
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(ret);
+    } else if (num_expired) {
+      // A goal expired!
+      std::array<uint8_t, 16> uuid;
+      for (size_t i = 0; i < 16; ++i) {
+        uuid[i] = expired_goals[0].goal_id.uuid[i];
+      }
+      pimpl_->goal_results_.erase(uuid);
+      pimpl_->result_requests_.erase(uuid);
+    }
+  }
+}
+
+void
+ServerBase::publish_status()
+{
+  rcl_ret_t ret;
+
+  // Get all goal handles known to C action server
+  rcl_action_goal_handle_t ** goal_handles = NULL;
+  size_t num_goals = 0;
+  ret = rcl_action_server_get_goal_handles(pimpl_->action_server_.get(), &goal_handles, &num_goals);
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+
+  auto status_msg = std::make_shared<action_msgs::msg::GoalStatusArray>();
+  status_msg->status_list.reserve(num_goals);
+  // Populate a c++ status message with the goals and their statuses
+  for (size_t i = 0; i < num_goals; ++i) {
+    rcl_action_goal_handle_t * goal_handle = goal_handles[i];
+    rcl_action_goal_info_t goal_info = rcl_action_get_zero_initialized_goal_info();
+    rcl_action_goal_state_t goal_state;
+
+    ret = rcl_action_goal_handle_get_info(goal_handle, &goal_info);
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(ret);
+    }
+
+    ret = rcl_action_goal_handle_get_status(goal_handle, &goal_state);
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(ret);
+    }
+
+    action_msgs::msg::GoalStatus msg;
+    msg.status = goal_state;
+    // Convert C goal info to C++ goal info
+    for (size_t i = 0; i < 16; ++i) {
+      msg.goal_info.goal_id.uuid[i] = goal_info.goal_id.uuid[i];
+    }
+    msg.goal_info.stamp.sec = goal_info.stamp.sec;
+    msg.goal_info.stamp.nanosec = goal_info.stamp.nanosec;
+
+    status_msg->status_list.push_back(msg);
+  }
+
+  // Publish the message through the status publisher
+  ret = rcl_action_publish_status(pimpl_->action_server_.get(), status_msg.get());
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+}
+
+void
+ServerBase::publish_result(const std::array<uint8_t, 16> & uuid, std::shared_ptr<void> result_msg)
+{
+  // Check that the goal exists
+  rcl_action_goal_info_t goal_info;
```

Comment:
Since the uuid -> `goal_info` conversion is happening in several places, it would be nice to wrap in a utility function. Perhaps after we have the GoalID wrapper introduced by #594

Reply:
Added conversion function in f717e9b5fcda8d36b3a7fb5d3557682b6bac745e